### PR TITLE
BTreeMap: Support custom allocators

### PR DESCRIPTION
**Pull request by exrook**
Moved to #98103

Related:
https://github.com/rust-lang/wg-allocators/issues/7
https://github.com/TimDiekmann/alloc-wg

Blocked on:
~~#77187~~
~~#78459~~ 
~~#95036~~

### CODE REVIEW

File: library/alloc/src/collections/btree/map.rs
Comment (TimDiekmann) [inline library/alloc/src/collections/btree/map.rs:598]: In this example, `new_in` should be used.

File: library/alloc/src/collections/btree/map.rs
Comment (TimDiekmann) [inline library/alloc/src/collections/btree/map.rs:127]: You added `#[unstable]` twice

```suggestion
pub struct BTreeMap<K, V, #[unstable(feature = "allocator_api", issue = "32838")] A: AllocRef = Global> {
```

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

**Comment by TimDiekmann** [comment]
cc @Amanieu

**Comment by Dylan-DPC** [comment]
unblocking this as both issues this was blocked on (as per the title) have been resolved

**Comment by JohnCSimon** [comment]
Triage: 
@Amanieu is this ready for review?

File: src/bootstrap/builder.rs
Comment (Amanieu) [inline src/bootstrap/builder.rs:1687]: Leftover from testing?

**Review by Amanieu — APPROVED** [review]
Sorry about the late review, this looks good modulo the leftover test code. Can you rebase to resolve the conflicts?

**Comment by TennyZhuang** [comment]
What's the status of the PR? I'd like to help with something.

**Comment by Amanieu** [comment]
@TennyZhuang Since the author is not responding, you can take over by forking their code and opening a new PR.

**Comment by TennyZhuang** [comment]
@exrook Sorry but I really need the feature, I'll take over the PR at the weekend. 🥰

**Comment by TennyZhuang** [comment]
@Amanieu Hi, I've resolved the conflicts and submitted a new PR #98015, PTAL Thanks!

**Comment by Dylan-DPC** [comment]
Closing this in favour of #98015

**Comment by exrook** [comment]
@TennyZhuang I'm glad to hear you're also looking to use this feature! My apologies for not responding sooner, I've been busy with other things, and when I last attempted a rebase on master back in May I was still hitting box-related ICEs so I've been working on a branch based on top of #95576, though it seems now those ICEs have been fixed in master :).

@Dylan-DPC Hi, could this be re-opened? I've put a lot of effort into keeping these changes in sync with master over the past year+ and a half and I'd like to see it through to being merged. :)

@Amanieu Thanks for the review! I've rebased on master and removed the leftover test code, so this should be good to go if you have no other comments

**Comment by TennyZhuang** [comment]
@exrook I’m glad to hear that, and really hope the PR can be merged ASAP.
@Dylan-DPC @Amanieu Can you help to reopen the PR and review it again? Sorry for the extra review work.

**Comment by Amanieu** [comment]
GIthub doesn't let me reopen the PR after a force-push/rebase. You'll need to open a new one.

**Comment by TennyZhuang** [comment]
@exrook I’ve found that it’s possible to reopen the PR, but it need you to revert the commits back to the one before closed, sorry.

See https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c

Feel free to reopen the PR or just create a new one.

**Comment by Dylan-DPC** [comment]
@exrook unfortunately you can't reöpen this. You can submit a new pull request and link this one in the description and we can take it from there

**Comment by exrook** [comment]
Thanks everyone! I've opened a new PR at #98103

**Pull request closed without merging**
